### PR TITLE
create a simple backlog to workaround ThrottlingException

### PR DESCRIPTION
**Pull request by ivanfmartinez**

### CODE REVIEW

File: src/main/java/ca/pjer/logback/AwsLogsAppender.java [hunk 1]
```diff
@@ -112,24 +115,39 @@ public synchronized void stop() {
...
+            final PutLogEventsResult result = awsLogs.putLogEvents(request);
             sequenceToken = result.getNextSequenceToken();
+            backLog.clear();
         } catch (DataAlreadyAcceptedException e) {
             sequenceToken = e.getExpectedSequenceToken();
         } catch (InvalidSequenceTokenException e) {
             sequenceToken = e.getExpectedSequenceToken();
             append(event);
+        } catch (AmazonServiceException e) {
+            backLog.add(inputLogEvent);
```
Comment (pierredavidbelanger) [inline src/main/java/ca/pjer/logback/AwsLogsAppender.java]: I wonder if adding to the backlog here when catching `AmazonServiceException` is not too agressive ?
What if `e` is not caused by a throttling problem .. we could end up adding infinitely to the backlog ?

Comment (pierredavidbelanger) [inline src/main/java/ca/pjer/logback/AwsLogsAppender.java]: I guess we could use `com.amazonaws.services.logs.model.LimitExceededException` ?

I actually have never saw a throttling error in my apps, so its only a guess.

Comment (ivanfmartinez) [inline src/main/java/ca/pjer/logback/AwsLogsAppender.java]: The Exception that I receive here is just the AmazonServiceException, the only way to check the Throthling is by checking the error Code String.

You are correct about other problems that can fill the backLog, maybe we can add a limit to the backLog, I'm using this logger in one web application and getting many times this exception, probably you can test in your environment creating a loop generating log messages to test.

The best way probably will be to add configuration for sending in batches. The problem That I see is some messages will be waiting until the buffer fill, and for applications with low logging some messages will not be sent to cloudwatch logs.

I did not see anything in logback that permit to create and send a batch of messages.

**Comment by robertoestivill** [comment]
Take a look at 

https://github.com/Virtual-Instruments/cloudwatch-log4j-appender/blob/master/src/main/java/com/vi/aws/logging/log4jappenders/CloudWatchAppender.java

It's using a queue and a background Thread for the dispatching.

**Comment by pierredavidbelanger** [comment]
Wow, I wanted this lib to stay simple.
I guess nothing is really simple ;)

The more I thing about this, and the more I think there is no way we can prevent the loss of some log event in the case of an abrupt JVM shutdown. Even persistence is not an option (just imagine running your app with something like `docker run -rm ... `.)

So, what about we keep the @ivanfmartinez `backLog` idea, and instead of a background thread (I do not like the idea of having yet an other thread to monitor), we just register a shutdown hook (`Runtime.getRuntime().addShutdownHook(...)`) to *try* to send the `backLog` on a best effort when the JVM is about to shutdown ?

**Comment by robertoestivill** [comment]
Hehe, yes I agree with keeping it simple :)

Shutdown hook sounds like a good idea for standalone jvm apps, but I do not think they will work on container servers. There, you will have to use a servlet context listener (or something like that).

Do you guys know what are the rules that aws applies when throttling the requests? per ip ? per user ?

**Comment by pierredavidbelanger** [comment]
My theory, I guess, is that: 

The applications that logs more frequently are the one that will face the ThrottlingException, so, they are the one that will see their backLog growing. But, since they log more frequently, they will also empty their backLog more frequently.

On the other side, applications that logs less frequently, will probably never face the ThrottlingException, so the backLog will never grow. But, in the rare case where this low logging application have something in the backLog, we will have the shutdown hook *safety* net.

**Comment by pierredavidbelanger** [comment]
@robertoestivill oh, indeed, you are absolutely right about web applications deployed in servlet containers. The backLog will just disappear.

**Comment by robertoestivill** [comment]
You can always provide a `ServiceContextListener` that does the same thing than the shutdown hook on the `contextDestroyed` method.

I haven't gone deep on how all the logback interfaces/implementations work, but how would you retrieve the appender instance from the Java code to accomplish this from the shutdown hook ?

**Comment by malkusch** [comment]
> I do not like the idea of having yet an other thread to monitor

I guess it's very common to use this implementation together with an `AsyncAppender`. Having a thread in this appender directly which would send cheap batches asynchronously would save having the thread of the `AsyncAppender`. IMO this would be a beneficial design.

**Comment by pierredavidbelanger** [comment]
@malkusch Very good point.

I guess the `AsyncAppender` would not even be required.

I will have a look at this very soon.

**Comment by pierredavidbelanger** [comment]
Ok guys, [here is a branch](https://github.com/pierredavidbelanger/logback-awslogs-appender/tree/async-mode) I pushed to play with some of your great ideas.

You can [get this version](https://github.com/pierredavidbelanger/logback-awslogs-appender/tree/async-mode#pomxml) from the sonatype.org snapshots repository:

```xml
<repositories>
  <repository>
    <id>sonatype-snapshots-repo</id>
    <url>https://oss.sonatype.org/content/repositories/snapshots</url>
    <releases><enabled>false</enabled></releases>
    <snapshots><enabled>true</enabled></snapshots>
  </repository>
</repositories>
```

You can check the [new configuration options](https://github.com/pierredavidbelanger/logback-awslogs-appender/tree/async-mode#more-configurations).

Also, it would be great if you can test this a little or [add more tests](https://github.com/pierredavidbelanger/logback-awslogs-appender/blob/async-mode/src/test/java/ca/pjer/logback/AsyncWorkerTest.java) ;)

Or, just leave me somme comments on the actual [AsyncWorker implementation](https://github.com/pierredavidbelanger/logback-awslogs-appender/commit/c03daae5ff1c37a7335cd64eac9c8e750bfacbd8#diff-39bf17be92ca1e63c37a32f88c71ddd0).

Enjoy!

**Comment by malkusch** [comment]
Thanks for the work. Here's a first feedback. If I would use the logger some concurrent web application (`ab -c 50` is enough) I see these exceptions:
> com.amazonaws.services.logs.model.InvalidParameterException: Log events in a single PutLogEvents request must be in chronological order. (Service: AWSLogs; Status Code: 400; Error Code: InvalidParameterException; Request ID: 5e80f7d6-adcd-11e6-80a6-b582835c364a)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleErrorResponse(AmazonHttpClient.java:1545)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeOneRequest(AmazonHttpClient.java:1183)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeHelper(AmazonHttpClient.java:964)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.doExecute(AmazonHttpClient.java:676)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeWithTimer(AmazonHttpClient.java:650)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.execute(AmazonHttpClient.java:633)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutor.access$300(AmazonHttpClient.java:601)
	at com.amazonaws.http.AmazonHttpClient$RequestExecutionBuilderImpl.execute(AmazonHttpClient.java:583)
	at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:447)
	at com.amazonaws.services.logs.AWSLogsClient.doInvoke(AWSLogsClient.java:1685)
	at com.amazonaws.services.logs.AWSLogsClient.invoke(AWSLogsClient.java:1661)
	at com.amazonaws.services.logs.AWSLogsClient.putLogEvents(AWSLogsClient.java:1402)
	at ca.pjer.logback.AWSLogsStub.logEvents(AWSLogsStub.java:58)
	at ca.pjer.logback.AsyncWorker.run(AsyncWorker.java:74)
	at java.lang.Thread.run(Thread.java:745)

I see if I can get more details about the circumstances.

**Comment by pierredavidbelanger** [comment]
@malkusch Wow, thank you!
Good catch. I was not aware of that AWS requirement.

**Comment by pierredavidbelanger** [comment]
@malkusch fixed [here](https://github.com/pierredavidbelanger/logback-awslogs-appender/commit/d7eb866fa98b9fa5fea2a254043d843cfeaa7c65) and deployed to sonatype snapshot repository as version [0.2.1-SNAPSHOT](https://github.com/pierredavidbelanger/logback-awslogs-appender/commit/4ea87e4a9f065138be0965e4fc2bacbdbc412dbe)

**Comment by pierredavidbelanger** [comment]
This was a patch that probably fixes the problem.

But, I fear that this is a side effect of thread contention, even with `ab -c 50`!

I think the contention happens [here](https://github.com/pierredavidbelanger/logback-awslogs-appender/blob/async-mode/src/main/java/ca/pjer/logback/AsyncWorker.java#L60) where many thread are trying to log at the same time, they wait for the monitor, possibly for a "long" time (the time it take to send log to AWS), then when the monitor is freed, threads wakes up in different order, thus adding into the list in the wrong order.

So, the bug you found may be a smell for a bad code design into the `AsyncWorker`. And I am open to discus about it.

**Comment by ivanfmartinez** [comment]
Maybe maintaining the queue in a Tree instead of a List can distribute the work for keeping they in order before sending and be more efficient.

The last patch with the sort will redo the sort if there are a problem sending the log items to AWS service, in the next retry.

**Comment by pierredavidbelanger** [comment]
@ivanfmartinez Indeed, you are right.

I like the idea of always keeping logs in order (thus distributing the the CPU time, and avoiding the double sort)

**Comment by pierredavidbelanger** [comment]
@ivanfmartinez commit 71a254c21ba501a9ac1f1976abfa3d686989a7a4 implements your idea

You will notice that I am using a `TreeSet`, because there is no such things as a `TreeList` (a `List` implementation that could stay sorted at insertion time) in Java.

Since `TreeSet` is a `Set`, one must keep in mind that trying to log more than once the same message at the exact same millisecond will result in only one of them being log (per batch size), because the will be `equals`.

This is not a bug, this is a flooding prevention feature ;)

**Comment by malkusch** [comment]
I tried the latest snapshot with `ab -c 500` and didn't see any concurrency issues. Well done!

**Comment by malkusch** [comment]
> Maybe maintaining the queue in a Tree instead of a List can distribute the work for keeping they in order before sending and be more efficient.

I think this is not true in respect of more efficient. The current implementation does now a relatively expensive insert into a tree and *block* all other application threads which want to append a log message as well. This introduces actually a contention point, compared to a cheap appending to a list and then just sorting in the logger thread.

This is of course just my opinion and I didn't back it up by an experiment.

**Comment by ivanfmartinez** [comment]
@pierredavidbelanger When I said to use a Tree I was thinking exactly in the TreeSet, because we have to keep the unique items ordered, and the flood prevention that is provided in the Set is good. :-)

@malkusch  I agree that (O(log(n))) is bigger than O(1), but in this case the n is small, we can consider it as O(1) because in the standard configuration the n will be less than 100. and the number of items in the Set will only increase if we lost the communication with CloudWatch service.

**Comment by pierredavidbelanger** [comment]
@ivanfmartinez @malkusch I do not know who is right. Without actual experimentation, this is hard to evaluate.

Anyways, [here is alternative implementation](https://github.com/pierredavidbelanger/logback-awslogs-appender/tree/async-mode-with-blocking-queue) (available as [0.2.3-SNAPSHOT](https://github.com/pierredavidbelanger/logback-awslogs-appender/tree/async-mode-with-blocking-queue#pomxml)) that do constant time and non blocking append. All [the burden](https://github.com/pierredavidbelanger/logback-awslogs-appender/blob/async-mode-with-blocking-queue/src/main/java/ca/pjer/logback/AWSLogsStub.java#L62) is now back on the worker thread.

I this this version is a trade off: less thread synchronisation for a potentially bigger memory pressure (on heavy logging applications).

I, personally, prefer this version.

But you guys tell me what should be the next version :)

**Comment by pierredavidbelanger** [comment]
I just realized that @ivanfmartinez 's original PR what about throttling exceptions and i did not even addressed that issue ;)

Comments are welcome.
I will work on this soon.

**Comment by pierredavidbelanger** [comment]
I wonder if this will not fix the throttling issue:

```java
ClientConfiguration clientConfiguration = PredefinedClientConfigurations.defaultConfig().withThrottledRetries(true);
AWSLogs awsLogs = new AWSLogsClient(clientConfiguration);
```

**Comment by ivanfmartinez** [comment]
@pierredavidbelanger  considering that the events are already sorted you can "break" here 
https://github.com/pierredavidbelanger/logback-awslogs-appender/blob/async-mode-with-blocking-queue/src/main/java/ca/pjer/logback/AWSLogsStub.java#L72
and  get lastTimestamp from the last item in the events collection. I think if we use ArrayBlockingQueue with a good capacity instead of LinkedBlockingQueue, this probably will be more efficient, because we can go directly to the end of queue to get lastTimestamp and in standard case without exceptions the queue size will be constant in ArrayBlockingQueue.

I think this solution are really better than the previous solutions.


About throttling I never seen this method before, but searching I just found that newer versions apparently has this enabled by default : https://aws.amazon.com/blogs/developer/throttled-retries-now-enabled-by-default/

**Comment by pierredavidbelanger** [comment]
@ivanfmartinez Interesting. 

I will add the `break`, that will save some CPU cycles.

Also, I was using `LinkedBlockingQueue`, because it does not mandate us to choose a fixed capacity. But, indeed `ArrayBlockingQueue` should be more efficient. But now: 
- what should be a *good* capacity ? :) I was thinking of `maxQueueSize * 3`, which will let the "client" threads insert up to 3 times our "batch" size.
- and, more importantly, what should we do when/if [queue.offer(..)](https://github.com/pierredavidbelanger/logback-awslogs-appender/blob/async-mode-with-blocking-queue/src/main/java/ca/pjer/logback/AsyncWorker.java#L64) will returns `false` ?

**Comment by ivanfmartinez** [comment]
@pierredavidbelanger we can use an configuration parameter for this capacity, when not specified *3 or *4 can be good numbers, considering that default queuesize is 50, with *4 the result will be 200 log entries.

its impossible to grow to infinite, maybe create a counter to indicate lost messages and when the queue is  accepting messages again add a message indicating the number of lost messages.

**Comment by malkusch** [comment]
Let's revive this, because sending in batches is a killer feature. I deployed 0.2.3-SNAPSHOT on a pet project and so far no issues. Also testing with `ab` looks fine.

> what should we do when/if queue.offer(..) will returns false ?

I can think of two options:
- Discard events
- Block

Both are valid approaches, which are actually implemented by `AsyncAppender`. There's a threshold when [unimportant events](https://github.com/qos-ch/logback/blob/v_1.2.3/logback-classic/src/main/java/ch/qos/logback/classic/AsyncAppender.java#L32-L40) are discarded. When it finally hits the limit it would [either discard the event or block](https://github.com/qos-ch/logback/blob/v_1.2.3/logback-core/src/main/java/ch/qos/logback/core/AsyncAppenderBase.java#L168-L174) (depending on the configuration).

**Comment by pierredavidbelanger** [comment]
Thank you @malkusch , this project needs a little dusting out :)

I myself do not use it anymore outside a non critical tool. I too use the 0.2.3-SNAPSHOT release for this.

I really like the links you provided.

I will do this.

**Comment by pierredavidbelanger** [comment]
This last commit 674b727736cba9f9831c55b986ae7fb3d53e5474 will definitely needs review and testing.

The 0.3.0-SNAPSHOT should be available in the [sonatype snapshot repo](https://github.com/pierredavidbelanger/logback-awslogs-appender/blob/674b727736cba9f9831c55b986ae7fb3d53e5474/README.md#pomxml).

Now, the [`append` method](https://github.com/pierredavidbelanger/logback-awslogs-appender/blob/674b727736cba9f9831c55b986ae7fb3d53e5474/src/main/java/ca/pjer/logback/AsyncWorker.java#L61) will now discard, block or not block depending on the queue `remainingCapacity` and the new [`maxBlockTimeMillis` config option](https://github.com/pierredavidbelanger/logback-awslogs-appender/blob/674b727736cba9f9831c55b986ae7fb3d53e5474/README.md#more-configurations).

**Comment by malkusch** [comment]
Wow, that was fast. I'll review that, but I'll be slower than you. Just a first note: If I understand correctly, you start discarding when the queue is full. The idea in AsyncAppender was discarding before the queue is full (discardingThreshold), so that there is a chance that the queue never gets actually full, effectively avoiding blocking.

**Comment by pierredavidbelanger** [comment]
@malkusch indeed , I start discarding only when the queue is full. This is intended because my `AsyncWorker` does not work exactly like the `AsyncAppender`.

My `AsyncWorker` buffers the events intentionally in the queue to be able to send them in batch (all at once when full), but the `AsyncAppender` only buffers them quickly and temporarily for the logger thread to keep up and send then somewhere one at a time at his own pace.

If I start discarding events when the queue is not full, I will always discard every DEBUG and INFO when the queue is passed the threshold, even if the buffering is the intended behaviour.

I am sorry, I hope it make sense ?

**Comment by malkusch** [comment]
Ok, I see. I would somehow feel better if the queue would have some additional buffer. Now there's always a chance that a new event comes in before the consumer thread could drain the queue.

Additionally I have one more thing to report from my 0.2.3-SNAPSHOT deployment:

    com.amazonaws.services.logs.model.InvalidParameterException: Log event too large: 267837 bytes exceeds limit of 262144 (Service: AWSLogs; Status Code: 400; Error Code: InvalidParameterException; Request ID: xxx)
            at com.amazonaws.http.AmazonHttpClient$RequestExecutor.handleErrorResponse(AmazonHttpClient.java:1545)
            at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeOneRequest(AmazonHttpClient.java:1183)
            at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeHelper(AmazonHttpClient.java:964)
            at com.amazonaws.http.AmazonHttpClient$RequestExecutor.doExecute(AmazonHttpClient.java:676)
            at com.amazonaws.http.AmazonHttpClient$RequestExecutor.executeWithTimer(AmazonHttpClient.java:650)
            at com.amazonaws.http.AmazonHttpClient$RequestExecutor.execute(AmazonHttpClient.java:633)
            at com.amazonaws.http.AmazonHttpClient$RequestExecutor.access$300(AmazonHttpClient.java:601)
            at com.amazonaws.http.AmazonHttpClient$RequestExecutionBuilderImpl.execute(AmazonHttpClient.java:583)
            at com.amazonaws.http.AmazonHttpClient.execute(AmazonHttpClient.java:447)
            at com.amazonaws.services.logs.AWSLogsClient.doInvoke(AWSLogsClient.java:1685)
            at com.amazonaws.services.logs.AWSLogsClient.invoke(AWSLogsClient.java:1661)
            at com.amazonaws.services.logs.AWSLogsClient.putLogEvents(AWSLogsClient.java:1402)
            at ca.pjer.logback.AWSLogsStub.logPreparedEvents(AWSLogsStub.java:86)
            at ca.pjer.logback.AWSLogsStub.logEvents(AWSLogsStub.java:75)
            at ca.pjer.logback.AsyncWorker.flush(AsyncWorker.java:93)
            at ca.pjer.logback.AsyncWorker.run(AsyncWorker.java:75)

Hmm, [the batch limit is 1MB](http://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_PutLogEvents.html). Reading that page makes me think we should enforce those limits (size, count) in this client as well. That message seems to indicate that a single event was too large. However I couldn't find anything in Amazon's documentation about such a limit. So I guess it's related to the batch size limit. I'll do some experiments and report back.

**Comment by pierredavidbelanger** [comment]
@malkusch 
> there's always a chance that a new event comes in before the consumer thread could drain the queue

You are absolutely right.

What if I rename the `maxQueueSize` for `maxBatchSize`; the queue will have a capacity of `maxBatchSize * 2` and, in addition to the current `maxFlushTimeMillis`, will drain (and start discarding unimportant events) when at `maxBatchSize`.

It will start discarding unimportant events  at `maxBatchSize` too.

So the logging thread will only start to block (using `maxBlockTimeMillis`) (or discard important events) at `maxBatchSize * 2`.

What do you think ?

**Comment by pierredavidbelanger** [comment]
Version 0.4.0-SNAPSHOT now contains the modifications I described in my last post.

**Comment by malkusch** [comment]
I think that's way better. However, I would still suggest starting discarding later. Somewhere after `maxBatchSize` and before the queue's limit. The reason is still the same, if consuming starts at `maxBatchSize` and discarding as well, there's still a chance that the consumer did not drain the queue. Adding a higher queue size post poned the point when blocking starts (which is good).

Just imaging the use case of two subsequent logger calls:

    logger.info("a");
    logger.debug("b");

When logging "a" results in adding the `maxBatchSize`th event, chances are that the consumer thread did not reach to draining the queue before logging "b", thus "b" got discarded. I think that's a not too rare situation.

I would suggest having three limits: maxBatchSize < discardThreshold < queue's capacity

Sorry for being so demanding.

**Comment by pierredavidbelanger** [comment]
@malkusch you are right again. I fail to see that one coming.

Should it be ok to have an hardcoded `discardThreshold` equals to `maxBatchSize  * 1.5` and a `maxQueueSize` equals to `maxBatchSize  * 2` ?

Or, should we let the user decide the `maxBatchSize`, `discardThreshold` and `maxQueueSize` ?

**Comment by malkusch** [comment]
Well, in my opinion best would be having those values configurable by the user but preset them with anything which makes sense (e.g. your suggested 1.5x and 2x batch size). Because if I would have to hardcode them, I lack experience to justify those values. I would preset them, so that users can use that appender with little configuration. I would still let them the option to configure them, in case their system behaves unexpected (e.g. too many dropped events during peaks).

However having said all that, **hardcoding is definetely OK** for now.

**Comment by malkusch** [comment]
I did add some review comments into the commit 44d1030c6b0a522210705b339226fe94d94b4117. I will do a few tests and come back again.

**Comment by pierredavidbelanger** [comment]
@malkusch Your insightful advices are really appreciated. Thank you.

**Comment by pierredavidbelanger** [comment]
I forgot to mention, version 1.0.0-SNAPSHOT is available on sonatype now.

**Comment by malkusch** [comment]
Thanks. This looks good for me. I also did a functional test which was fine. 

To make it complete, let's talk about [AWS' constraints](http://docs.aws.amazon.com/AmazonCloudWatchLogs/latest/APIReference/API_PutLogEvents.html):
- The maximum batch size is 1,048,576 bytes
- The maximum number of log events in a batch is 10,000.

Let's throw an IllegalArgumentException in [`setMaxBatchSize()`](https://github.com/pierredavidbelanger/logback-awslogs-appender/blob/async-mode-with-blocking-queue/src/main/java/ca/pjer/logback/AwsLogsAppender.java#L72-L74) if the batch size is not within 1 and 10000.

What about the batch size? Requeueing (or discarding) the offending events and adding a warning to decrease the batch size might be an option.

As a side note: I did the functional test as a JUnit test and noticed that `AwsLogsAppender.stop()` was not called, which leads to a few unlogged events. However I guess that's surely not an issue of this appender and most likely me not using Logback correctly.

**Comment by pierredavidbelanger** [comment]
Thank you for your feedbacks !

I will consider this issue closed and finally merge this branch into master.

> To make it complete, let's talk about AWS' constraints:

And I will address those in issue #6 that you opened a weak ago (thank you !).

> AwsLogsAppender.stop() was not called, which leads to a few unlogged events

Yeah, this is possible.

Have a look [here](https://github.com/pierredavidbelanger/logback-awslogs-appender/tree/async-mode-with-blocking-queue#more-configurations), where I talk about :

```xml
...
    <!-- Register the shutdown hook to allow logback to cleanly stop appenders -->
    <!-- this is strongly recommend when using AwsLogsAppender in async mode, -->
    <!-- to allow the queue to flush on exit -->
    <shutdownHook class="ch.qos.logback.core.hook.DelayingShutdownHook"/>
...
```

Try this and open a new issue if it does not work.

**Pull request closed without merging**